stretch_core/stretch_core/stretch_driver.py

```diff
@@ -456,152 +456,159 @@ def runstop_service_callback(self, request):
-                                                             Trigger,
-                                                             self.position_mode_service_callback)
-        
-        self.stop_the_robot_service = rospy.Service('/stop_the_robot',
-                                                    Trigger,
-                                                    self.stop_the_robot_callback)
-
-        self.calibrate_the_robot_service = rospy.Service('/calibrate_the_robot',
-                                                         Trigger,
-                                                         self.calibrate_callback)
-
-        self.runstop_service = rospy.Service('/runstop',
-                                              SetBool,
-                                              self.runstop_service_callback)
-
-        try:
-            # start loop to command the mobile base velocity, publish
-            # odometry, and publish joint states
-            while not rospy.is_shutdown():
-                self.command_mobile_base_velocity_and_publish_state()
-                command_base_velocity_and_publish_joint_state_rate.sleep()
-        except (rospy.ROSInterruptException, ThreadServiceExit):
-            self.robot.stop()
-            rospy.signal_shutdown("stretch_driver shutdown")
+        self.switch_to_navigation_mode_service = self.create_service(Trigger,
+                                                                     '/switch_to_navigation_mode',
+                                                                     self.navigation_mode_service_callback)
+
+        self.switch_to_position_mode_service = self.create_service(Trigger,
+                                                                   '/switch_to_position_mode',
+                                                                   self.position_mode_service_callback)
+
+        self.stop_the_robot_service = self.create_service(Trigger,
+                                                          '/stop_the_robot',
+                                                          self.stop_the_robot_callback)
+
+        self.calibrate_the_robot_service = self.create_service(Trigger,
+                                                               '/calibrate_the_robot',
+                                                               self.calibrate_callback)
+
+        self.runstop_service = self.create_service(SetBool,
+                                                   '/runstop',
+                                                   self.runstop_service_callback)
+
+        # start loop to command the mobile base velocity, publish
+        # odometry, and publish joint states
+        timer_period = 1.0 / self.joint_state_rate
+        self.timer = self.create_timer(timer_period, self.command_mobile_base_velocity_and_publish_state)
+
+def main():
+    try:
+        rclpy.init()
+        executor = MultiThreadedExecutor(num_threads=2)
```

Comment:
If I recall correctly, in ROS 1 each publisher/subscriber/action client [creates its own thread.](https://nu-msr.github.io/me495_site/lecture08_threads.html). Here, we do want multiple threads because otherwise if you are executing a long action (like a trajectory) it will mean that your other callbacks (the timer, other subscriptions) won't fire.